src/Build/BackEnd/TaskExecutionHost/TaskExecutionHost.cs

```diff
@@ -1283,10 +1283,13 @@ private bool InitializeTaskVectorParameter
         /// </remarks>
         private bool InternalSetTaskParameter(TaskPropertyInfo parameter, IList parameterValue)
         {
-            if (LogTaskInputs && !_taskLoggingContext.LoggingService.OnlyLogCriticalEvents && parameterValue.Count > 0)
+            if (LogTaskInputs &&
```

Comment:
Rather than threading parameter through as you did here, could we instead add a (private) set method to LogTaskInputs? The only potential drawback I see is that that would set it for a TaskExecutionHost rather than setting it for a particular task, but you could presumably add something that changes it and reverts it to its previous value afterwards, and that would still consolidate this logic in one place.

If you want to add the option to log inputs but not their metadata, you could also change LogTaskInputs into an enum with options like `LogAllInputs`, `LogInputsWithoutMetadata`, and `DoNotLogInputs`. Since TaskExecutionHost is internal, I don't think that's a breaking change.

Reply:
Yes, I think LogTaskInputs needs reworking. It has various issues, e.g. users have no control over it because it's coupled with verbosity. 

I didn't quite understand which set method you're suggesting to add and where? I think we want to control whether to log item metadata per task.

Reply:
Looking at this again, I think my method wouldn't work because it would change logging behavior for all tasks executing in parallel on a TaskExecutionHost. We could presumably make separate TaskExecutionHosts, each of which has a unique value for LogTaskInputs and never executes a task that doesn't agree with their version, but that would be a heavyweight solution.

Although I no longer think it's a great solution, here's what I was thinking:
protected set method [here](https://github.com/microsoft/msbuild/blob/master/src/Build/BackEnd/TaskExecutionHost/AddInParts/ITaskExecutionHost.cs#L58)
Choose a couple key places (like [here](https://github.com/microsoft/msbuild/blob/master/src/Build/BackEnd/TaskExecutionHost/TaskExecutionHost.cs#L429), for instance) where you can set LogTaskInputs to a new value (with LogTaskInputs being an enum including whether to log inputs and whether to log metadata). Then you can reset it in a finally block [here](https://github.com/microsoft/msbuild/blob/master/src/Build/BackEnd/TaskExecutionHost/TaskExecutionHost.cs#L510). If there's only one thread running on a TaskExecutionHost at a time, or they all have the same logging behavior, that would mean you wouldn't have to thread it through so many function calls.